projects/packages/licensing/tests/php/class-test-licensing-endpoints.php

```diff
@@ -0,0 +1,161 @@
+<?php
+/**
+ * Tests the endpoints within the Licensing package.
+ *
+ * @package automattic/jetpack-licensing
+ */
+
+namespace Automattic\Jetpack;
+
+use WorDBless\BaseTestCase;
+
+/**
+ * Class Test_Licensing
+ *
+ * @package Automattic\Jetpack
+ */
+class Test_Licensing_Endpoints extends BaseTestCase {
+
+	/**
+	 * Used to store an instance of the WP_REST_Server.
+	 *
+	 * @var WP_REST_Server
+	 */
+	private $server;
+
+	/**
+	 * Ensure that Licensing package has been setup and hooks have been registered.
+	 */
+	public static function setUpBeforeClass() : void { // phpcs:ignore.
```

Comment:
Since Jetpack still supports PHP 5.6+ to match WordPress Core requirements, you won't be able to use return types (introduced in PHP 7.1 I think). I would recommend adding a `@return` to the docblock instead. Not perfect, but PHP 5.6 friendly :)

Reply:
The issue with not having the return types is this:

```
Fatal error: Declaration of Automattic\Jetpack\Test_Licensing_Endpoints::setUpBeforeClass() must be compatible with PHPUnit\Framework\TestCase::setUpBeforeClass(): void in ...jetpack/projects/packages/licensing/tests/php/class-test-licensing-endpoints.php on line 29
```

The issue here is that PHPUnit itself has set the return type on the parent methods. Since we're in the tests here, as opposed to the plugin files, will it be an issue to have the return type?

Reply:
Hm, my bad, I didn't notice that. I tried an alternative to get rid of those methods with something like this, but I could not get it to work:

```diff
diff --git a/projects/packages/licensing/tests/php/class-test-licensing-endpoints.php b/projects/packages/licensing/tests/php/class-test-licensing-endpoints.php
index 414d655225..5b7c040d6d 100644
--- a/projects/packages/licensing/tests/php/class-test-licensing-endpoints.php
+++ b/projects/packages/licensing/tests/php/class-test-licensing-endpoints.php
@@ -7,15 +7,18 @@
 
 namespace Automattic\Jetpack;
 
-use WorDBless\BaseTestCase;
+use Automattic\Jetpack\Licensing\Endpoints;
+
+use PHPUnit\Framework\TestCase;
+use WP_REST_Request;
+use WP_REST_Server;
 
 /**
  * Class Test_Licensing
  *
  * @package Automattic\Jetpack
  */
-class Test_Licensing_Endpoints extends BaseTestCase {
-
+class Test_Licensing_Endpoints extends TestCase {
 	/**
 	 * Used to store an instance of the WP_REST_Server.
 	 *
@@ -24,36 +27,39 @@ class Test_Licensing_Endpoints extends BaseTestCase {
 	private $server;
 
 	/**
-	 * Ensure that Licensing package has been setup and hooks have been registered.
+	 * Used to store an instance of Endpoints.
+	 *
+	 * @var Endpoints
 	 */
-	public static function setUpBeforeClass() : void { // phpcs:ignore.
-		parent::setUpBeforeClass();
-
-		$licensing = new Licensing();
-		$licensing->initialize();
-		do_action( 'rest_api_init' );
-	}
+	protected $endpoints;
 
 	/**
 	 * Setup environment for REST API endpoints test.
 	 */
-	public function setUp(): void { // phpcs:ignore.
-		parent::set_up_wordbless();
-
+	public function set_up() {
+		parent::set_up();
 		global $wp_rest_server;
-		$wp_rest_server = new \WP_REST_Server();
-		$this->server   = $wp_rest_server;
+
+		$wp_rest_server  = new WP_REST_Server();
+		$this->server    = $wp_rest_server;
+		$this->endpoints = new Endpoints();
+
+		// Register REST routes.
+		add_action( 'rest_api_init', array( $this->endpoints, 'initialize_endpoints' ) );
+
 		do_action( 'rest_api_init' );
 	}
 
 	/**
 	 * Clean environment for REST API endpoints test.
 	 */
-	public function tearDown(): void { // phpcs:ignore.
-		parent::tear_down_wordbless();
+	public function tear_down() {
+		parent::tear_down();
 
 		global $wp_rest_server;
 		$wp_rest_server = null;
+
+		remove_action( 'rest_api_init', array( $this->endpoints, 'initialize_endpoints' ) );
 	}
 
 	/**
@@ -97,7 +103,7 @@ class Test_Licensing_Endpoints extends BaseTestCase {
 	 * @return WP_REST_Response
 	 */
 	protected function create_and_get_request( $route = '', $json_params = array(), $method = 'GET', $params = array() ) {
-		$request = new \WP_REST_Request( $method, "/jetpack/v4/$route" );
+		$request = new WP_REST_Request( $method, "/jetpack/v4/$route" );
 
 		if ( 'GET' !== $method && ! empty( $json_params ) ) {
 			$request->set_header( 'content-type', 'application/json' );
@@ -137,7 +143,7 @@ class Test_Licensing_Endpoints extends BaseTestCase {
 		wp_set_current_user( $user->ID );
 
 		// Should be empty by default.
-		$request  = new \WP_REST_Request( 'GET', '/jetpack/v4/licensing/error' );
+		$request  = new WP_REST_Request( 'GET', '/jetpack/v4/licensing/error' );
 		$response = $this->server->dispatch( $request );
 		$this->assertResponseStatus( 200, $response );
 		$this->assertSame( '', $response->get_data() );
@@ -154,7 +160,7 @@ class Test_Licensing_Endpoints extends BaseTestCase {
 		$this->assertTrue( $response->get_data() );
 
 		// Should return updated value.
-		$request  = new \WP_REST_Request( 'GET', '/jetpack/v4/licensing/error' );
+		$request  = new WP_REST_Request( 'GET', '/jetpack/v4/licensing/error' );
 		$response = $this->server->dispatch( $request );
 		$this->assertResponseStatus( 200, $response );
 		$this->assertEquals( 'foo', $response->get_data() );

```

I still get an error when running the tests, so I must be missing something here:

```
There was 1 error:

1) Automattic\Jetpack\Test_Licensing_Endpoints::test_licensing_error
Error: Call to a member function dispatch() on null

/Users/jeherve/code/jetpack/projects/packages/licensing/tests/php/class-test-licensing-endpoints.php:147
```

> Since we're in the tests here, as opposed to the plugin files, will it be an issue to have the return type and for the tests to not be compatible with 5.6?

I'm afraid the monorepo isn't set up for that right now, since we run all tests in different PHP environments. Everything must be PHP 5.6 compatible right now.

I also had to rebase your branch, I pushed an unrelated commit to it just now while working on another thing at the same time. Sorry about that! You'll need to pull a fresh copy of that branch when you start working I'm afraid.

Reply:
I've pushed a change that works around the test setup so that we don't hit the 5.6 lint issue for now. I'd rather be able to use the standard `setUp`, `tearDown`, `setUpBeforeClass`, but this is fine for now.

Reply:
We have the Yoast phpunit polyfills. Would this be fine? https://github.com/Yoast/PHPUnit-Polyfills#testcases

Reply:
Realize we're also trying to use WorDBless so opening https://github.com/Automattic/wordbless/issues/52